examples/static_graph_optimizations/mnist/train_mnist.py

```diff
@@ -0,0 +1,149 @@
+"""MNIST example with static subgraph optimizations.
+
+This is a version of the Chainer MNIST example that has been modified
+to support the static subgraph optimizations feature. Note that
+the code is mostly unchanged except for the addition of the
+`@static_graph` decorator to the model chain's `__call__()` method.
+"""
+from __future__ import print_function
+
+import argparse
+
+import chainer
+import chainer.functions as F
+from chainer.graph_optimizations.static_graph import static_graph
+from chainer.graph_optimizations.static_graph_utilities \
+    import static_code
+import chainer.links as L
+from chainer import training
+from chainer.training import extensions
+
+
+# Network definition
+class MLP(chainer.Chain):
+
+    def __init__(self, n_units, n_out):
+        super(MLP, self).__init__()
+        with self.init_scope():
+            # the size of the inputs to each layer will be inferred
+            self.l1 = L.Linear(None, n_units)  # n_in -> n_units
+            self.l2 = L.Linear(None, n_units)  # n_units -> n_units
+            self.l3 = L.Linear(None, n_out)  # n_units -> n_out
+        self.side_effect_counter = 0
+
+    @static_code
+    def example_side_effect(self):
```

Comment:
Honestly, I had the impression that this code is a little complicated when I first saw this one.
An example of a side effect code is useful, but I think writing a side effect code in a forward computation like this is relatively a rare situation.
I feel that removing the side effect code will make the example simpler than now. It might be good for casual users.
The current example may be put in another place. How do you think about this?

Reply:
I agree that it would be rare to write side effect code like this. I added the side effect code to the example after other feedback, so I think it would be good to keep an example somewhere. But I also think it could be better to move it to another place or at least mention in the example that such usage may not be common. Other ideas would be to document an example in a "design notes" document or to just include a test case for it.

Reply:
For now, I have decided to modify the example so that side effect code is not used by default. However, I have added a second MLP chain to the MNIST example that uses side effect code that the user can optionally enable. I have also updated the documentation to mention the usage of side effect code is expected to be rare but an example is included for completeness. Does it seem okay?